pkg/operator/configobservation/authentication/apiaudiences.go

```diff
@@ -0,0 +1,112 @@
+package auth
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"k8s.io/klog/v2"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	"github.com/openshift/cluster-authentication-operator/pkg/operator/configobservation"
+)
+
+var (
+	audiencesPath = []string{"apiServerArguments", "api-audiences"}
+)
+
+// ObserveAPIAudiences changes apiServerArguments.api-audiences from
+// the default value if Authentication.Spec.ServiceAccountIssuer specifies a valid
+// non-empty value.
+func ObserveAPIAudiences(
+	genericListers configobserver.Listers,
+	recorder events.Recorder,
+	existingConfig map[string]interface{},
+) (ret map[string]interface{}, errs []error) {
+	defer func() {
+		ret = configobserver.Pruned(ret, audiencesPath)
+	}()
+
+	listers := genericListers.(configobservation.Listers)
+	errs = []error{}
+	var issuerChanged bool
+	var existingAudience, newAudience string
+	// when the issuer will change, indicate that by setting `issuerChanged` to true
+	// to emit the informative event
+	defer func() {
+		if !issuerChanged {
+			return
+		}
+
+		recorder.Eventf(
+			"ObserveAPIAudiences",
+			"service account issuer changed from %s to %s",
+			existingAudience, newAudience,
+		)
+	}()
+
+	existingAudiences, _, err := unstructured.NestedStringSlice(existingConfig, audiencesPath...)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("unable to extract service account issuer from unstructured: %v", err))
+	}
+
+	// we're using the value of authentication.spec.issuer which is always a string, not a slice
+	if len(existingAudiences) > 0 {
```

Comment:
it's confusing that the configuration field is plural but only single-value in logic because it correlates with single value service account issuer.

does it make sense to emit some "quirky configuration" back to the user, i.e. setting a Status field with InvalidConfiguration=True?

Reply:
i think we have the same logic in KAS-o though.

Reply:
otoh this config should never be set by the user (sorry for thinking while commenting).

Reply:
This is an internal configuration that's not exposed to the user so I wouldn't worry about it too much. There would only be multiple values here if someone or something was touching something they aren't supposed to, and it would be rewritten right away.